tests/vcs/git/test_backend.py

```diff
@@ -75,3 +125,49 @@ def test_urlpathjoin(url: str, expected_result: str) -> None:
     path = "../other-repo"
     result = urlpathjoin(url, path)
     assert result == expected_result
+
+
+def test_resolve() -> None:
+    git_ref = GitRefSpec("main", "1234", "v2")
+
+    assert git_ref.branch == "main"
+    assert git_ref.revision == "1234"
+    assert git_ref.tag == "v2"
+    assert git_ref.ref == b"HEAD"
+
+
+def test_git_ref_spec_resolve_branch(fetch_pack_result: FetchPackResult) -> None:
+    mock_fetch_pack_result = fetch_pack_result
+
+    refspec = GitRefSpec(branch="main")
+    refspec.resolve(mock_fetch_pack_result)
+
+    assert refspec.ref == b"refs/heads/main"
+    assert refspec.branch == "main"
+    assert refspec.revision is None
+    assert refspec.tag is None
+
+
+def test_git_ref_spec_resolve_tag(fetch_pack_result: FetchPackResult) -> None:
+    mock_fetch_pack_result = fetch_pack_result
+
+    refspec = GitRefSpec(revision="v1.0.0")
+    refspec.resolve(mock_fetch_pack_result)
+
+    assert refspec.ref == annotated_tag(b"refs/tags/v1.0.0")
+    assert refspec.branch is None
+    assert refspec.revision is None
+    assert refspec.tag == "v1.0.0"
+
+
+def test_git_ref_spec_resolve_sha(fetch_pack_result: FetchPackResult) -> None:
+    mock_fetch_pack_result = fetch_pack_result
+
+    refspec = GitRefSpec(revision="abc")
+
+    refspec.resolve(mock_fetch_pack_result)
+
+    assert refspec.ref == b"refs/heads/main"
+    assert refspec.branch is None
+    assert refspec.tag is None
+    assert refspec.revision == "abc"
```

Comment:
Can we merge these three `ref_spec_resolve` tests into one parameterized test?

There are still some uncovered branches in `_normalise`, e.g. specifying a `branch` as `revision`, which can be covered easily. If we add more test variants, a parameterized tests will make even more sense.